sound/soc/sof/topology.c

```diff
@@ -2214,28 +2224,52 @@ static int sof_link_hda_process(struct snd_sof_dev *sdev,
 		if (!dai->name)
 			continue;
 
-		if (strcmp(link->name, dai->name) == 0 &&
-		    dai->comp_dai.direction == direction) {
+		if (strcmp(link->name, dai->name) == 0) {
+			if (dai->comp_dai.direction ==
+			    SNDRV_PCM_STREAM_PLAYBACK) {
+				config->hda.link_dma_ch = tx_slot;
+				playback_found = 1;
+			} else {
+				config->hda.link_dma_ch = rx_slot;
+				capture_found = 1;
+			}
+
 			config->dai_index = dai->comp_dai.dai_index;
-			found = 1;
-			break;
+
+			/* save config in dai component */
+			dai->dai_config = kmemdup(config, size, GFP_KERNEL);
+			if (!dai->dai_config)
+				return -ENOMEM;
+
+			/* send message to DSP */
+			ret = sof_ipc_tx_message(sdev->ipc,
+						 config->hdr.cmd, config, size,
+						 &reply, sizeof(reply));
+
+			if (ret < 0) {
+				dev_err(sdev->dev, "error: failed to set DAI config for direction:%d of HDA dai %d\n",
+					dai->comp_dai.direction,
+					config->dai_index);
+
+				return ret;
+			}
 		}
 	}
 
-	if (!found) {
-		dev_err(sdev->dev, "error: failed to find dai %s in %s",
-			link->name, __func__);
+	if (link->dpcm_playback && !playback_found) {
+		dev_err(sdev->dev, "error: failed to find playback dai for hda dai link %s",
+			link->name);
+
 		return -EINVAL;
 	}
 
-	config->hda.link_dma_ch = slot;
-
-	/* send message to DSP */
-	ret = sof_ipc_tx_message(sdev->ipc,
-				 config->hdr.cmd, config, size, &reply,
-				 sizeof(reply));
+	if (link->dpcm_capture && !capture_found) {
+		dev_err(sdev->dev, "error: failed to find capture dai for hda dai link %s",
+			link->name);
+		return -EINVAL;
+	}
 
```

Comment:
@RanderWang as we discussed, let's change this to be:

```
if ((link->dpcm_playback && playback_found) ||
    (link->dpcm_capture && capture_found)) {
        dev_info(...);//success
} else {
        dev_err();// error.
        return -EINVAL;
}
```

Reply:
@RanderWang @keyonjie should it be a warning instead of error. My point is that nothing would be broken in this case. Its just that we will not have a pcm that uses the BE with this config in this case.

Reply:
@ranj063 yes, I get your idea and I am refining it now